### PR TITLE
SSO service documentation

### LINKED ISSUE #2955: How Auth works doc
After today's meeting turned out Auth is complicated and keeps getting explained every time again and again. There needs to be a single place to look for that information - the doc.

I have started a scrap one in Wiki - https://github.com/Seneca-CDOT/telescope/wiki/How-Auth-works
To save time and not wait for PRs on small fixes, we instead can work on it in the Wiki, until we have a cohesive result.

#### Here are some sources where you can gather information on it:
Screen Shot 2021-02-19 at 7 11 49 PM

[Blog post](https://blog.humphd.org/not-so-simple-saml)

If you are familiar with our Auth process, please add what you know to the page!

### PR DESCRIPTION
**Pull request by sirinoks**
## Issue This PR Addresses

#3291 Only the sso part 

Possibly, 
Fixes #2955 

However, for that specific issue I would want to describe SAML-2 part in a separate document, or add SAML-2 information to this doc later. 

## Type of Change

- [x] **Documentation Update**: Change which improves documentation

## Description

Describes the step by step process of auth. It is the first draft and is open to edits.

## Steps to test the PR

- `cd src/web/docusaurus`
- `pnpm start`
- localhost:4631
- Click on Docs tab
- Click on API services, then `Authentication and Authorization`
- Read the content. Does it make sense? Can it be better?

## Checklist

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [x] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

No tests for just docs. Prettier was okay with this.
No need for screenshots for just text.

<hr />
Might need to be updated with Supabase information and whatever is in #3030 . If this gets merged, please make sure to add an issue to describe #3030 in documentation somewhere.

### CODE REVIEW

**Comment by sirinoks** [comment]
All related info: 
- Issue [How auth works doc](https://github.com/Seneca-CDOT/telescope/issues/2955)
- [Draft of the doc mentioned on wiki](https://github.com/Seneca-CDOT/telescope/wiki/How-Auth-works)
- [The recording where SSO process is described](https://github.com/Seneca-CDOT/telescope/issues/2955#issuecomment-1096686448)
- Issue that [tracks all API documentation tasks](https://github.com/Seneca-CDOT/telescope/issues/3291)
- Issue that [wants general API documentation](https://github.com/Seneca-CDOT/telescope/issues/203)

_Things that might need to be added later:_
- Issue of [some new SSO functionality that was not yet documented](https://github.com/Seneca-CDOT/telescope/issues/3030)
- [SAML-2 blog post](https://blog.humphd.org/not-so-simple-saml/)

File: src/web/docusaurus/docs/api-services/auth.md [hunk 1]
```diff
@@ -1,3 +1,86 @@
+---
+sidebar_position: 5
+---
+
 # Authentication and Authorization
 
-Documentation in construction...
+## Summary
+
+SSO service is required for authentication. Telescope doesn't store user information itself, we just send it to Seneca's SSO. From that interaction, Telescope creates a token that will only work for a limited amount of time. The token is used in authenticating in various Telescope services, without a need to make requests to Seneca's servers every time. The same token system is also used throughout our Supabase back end.
```
Comment (JiaHua-Zou) [inline src/web/docusaurus/docs/api-services/auth.md:9]: Maybe we can reword it from:
`The token is used in authenticating in various Telescope services, without a need to make requests to Seneca's servers every time` 
 to -> `The token is used to authenticate various Telescope services without sending requests to Seneca's server every time.`

Change `back end`  -> `backend`

File: src/web/docusaurus/docs/api-services/auth.md [hunk 2]
```diff
@@ -1,3 +1,86 @@
...
+
+## Stages
+
+_This is what happens step by step in the SSO service:_
+
+- Request to log in is sent with the user entering their log in information into Telescope's front end.
+- Log in data is sent to Seneca's servers.
```
Comment (JiaHua-Zou) [inline src/web/docusaurus/docs/api-services/auth.md:16]: `Log in` - > `Login`

File: src/web/docusaurus/docs/api-services/auth.md [hunk 3]
```diff
@@ -1,3 +1,86 @@
...
-Documentation in construction...
+## Summary
+
+SSO service is required for authentication. Telescope doesn't store user information itself, we just send it to Seneca's SSO. From that interaction, Telescope creates a token that will only work for a limited amount of time. The token is used in authenticating in various Telescope services, without a need to make requests to Seneca's servers every time. The same token system is also used throughout our Supabase back end.
+
+## Stages
+
+_This is what happens step by step in the SSO service:_
+
+- Request to log in is sent with the user entering their log in information into Telescope's front end.
```
Comment (JiaHua-Zou) [inline src/web/docusaurus/docs/api-services/auth.md:15]: Maybe we can reword it to: `The action of logging in to the Telescope website will send a request to login.`

File: src/web/docusaurus/docs/api-services/auth.md [hunk 4]
```diff
@@ -1,3 +1,86 @@
...
+   :::
+
+## Proceeding stages
+
+- SSO service signs the token using a JWT secret, making it viable for a limited amount of time.
+
+:::info
+
+JWT is a string that has 3 parts. These parts are JavaScript objects made into a single line.
+Each part then base 64 encoded and is meant to be put in a URL. Parts are divided with `.` (dots)
```
Comment (JiaHua-Zou) [inline src/web/docusaurus/docs/api-services/auth.md:69]: I don't quite understand this. The 3 parts are encoded with the base64 encoder?

File: src/web/docusaurus/docs/api-services/auth.md [hunk 5]
```diff
@@ -1,3 +1,86 @@
...
+- Microsoft SAML-2.
+
+  :::info
+
+  See more in [Not so simple SAML](https://blog.humphd.org/not-so-simple-saml/)
+
+  :::
+
+- Auth cookies are returned to SSO service.
+- SSO service creates a token. A token is basically JSON data.
```
Comment (tcvan0707) [inline src/web/docusaurus/docs/api-services/auth.md:26]:
```suggestion
- SSO service creates a token, which is basically a JSON data.
```

File: src/web/docusaurus/docs/api-services/auth.md [hunk 6]
```diff
@@ -1,3 +1,86 @@
...
+
+### Tokens
+
+_Token includes information such as:_
+
+1. Who issued it (Us - Telescope server)
+2. Who the audience is (For whom it is made and works for. So, Telescope servers)
+3. Who is the subject of the token (Some ID. In our case, a hashed user email)
```
Comment (tcvan0707) [inline src/web/docusaurus/docs/api-services/auth.md:34]:
```suggestion
3. Who is the subject of the token (Some IDs. In our case, a hashed user email)
```

File: src/web/docusaurus/docs/api-services/auth.md [hunk 7]
```diff
@@ -1,3 +1,86 @@
...
+4. Claims. Data about the user.
+
+### Claims
+
+_Claims include following data:_
+
+1. Email
+2. First name
+3. Last name
+4. Display name (composite of first and last names)
```
Comment (tcvan0707) [inline src/web/docusaurus/docs/api-services/auth.md:44]:
```suggestion
4. Display name (consists of first and last names)
```

**Review by tcvan0707 — CHANGES_REQUESTED** [review]

File: src/web/docusaurus/docs/api-services/auth.md (hunk 4 above)
Comment (sirinoks) [reply src/web/docusaurus/docs/api-services/auth.md:69]: The information is encoded with base64. This is what was said in the [meeting](https://seneca.sharepoint.com/sites/Telescope/Shared%20Documents/Forms/AllItems.aspx?id=%2Fsites%2FTelescope%2FShared%20Documents%2FGeneral%2FRecordings%2FJWT%20Secret%20in%20Telescope%2D20220406%5F135311%2DMeeting%20Recording%2Emp4&parent=%2Fsites%2FTelescope%2FShared%20Documents%2FGeneral%2FRecordings), specifically at 08:01.

File: src/web/docusaurus/docs/api-services/auth.md (hunk 3 above)
Comment (sirinoks) [reply src/web/docusaurus/docs/api-services/auth.md]: Sure. I think your suggestion could also be rewritten better, but it is already a step up to the previous version.

File: src/web/docusaurus/docs/api-services/auth.md (hunk 4 above)
Comment (sirinoks) [reply src/web/docusaurus/docs/api-services/auth.md:69]: If you have a suggestion how to reword this better, make another comment please

**Review by tcvan0707 — DISMISSED** [review]
Looks good to me!

File: src/web/docusaurus/docs/api-services/auth.md (hunk 4 above)
Comment (JiaHua-Zou) [reply src/web/docusaurus/docs/api-services/auth.md:69]: We can reword it to
` Each parts are encoded with the base64 encoder and then it is used in a URL`

Comment (JiaHua-Zou) [reply src/web/docusaurus/docs/api-services/auth.md:69]: We can reword it to
` Each parts are encoded with the base64 encoder and then it is used in a URL`

**Review by JiaHua-Zou — CHANGES_REQUESTED** [review]
We can reword it to
` Each parts are encoded with the base64 encoder and then it is used in a URL`

File: src/web/docusaurus/docs/api-services/auth.md (hunk 4 above)
Comment (sirinoks) [reply src/web/docusaurus/docs/api-services/auth.md:69]: > We can reword it to ` Each parts are encoded with the base64 encoder and then it is used in a URL`

I'm not sure this wording is better, honestly

File: src/web/docusaurus/docs/api-services/auth.md [hunk 8]
```diff
@@ -1,3 +1,86 @@
...
+- Microsoft SAML-2.
+
+  :::info
+
+  See more in [Not so simple SAML](https://blog.humphd.org/not-so-simple-saml/)
+
+  :::
+
+- Auth cookies are returned to SSO service.
+- SSO service creates a token, which is basically a JSON data.
```
Comment (TueeNguyen) [inline src/web/docusaurus/docs/api-services/auth.md:26]:
```suggestion
SSO service creates a JWT token, which is basically encrypted JSON data.
```

File: src/web/docusaurus/docs/api-services/auth.md [hunk 9]
```diff
@@ -1,3 +1,86 @@
...
+
+- SSO service signs the token using a JWT secret, making it viable for a limited amount of time.
+
+:::info
+
+JWT is a string that has 3 parts. These parts are JavaScript objects made into a single line.
+Each part then base 64 encoded and is meant to be put in a URL. Parts are divided with `.` (dots)
+In the end, there is a secret which is used to create a cryptographic hash.
+You can't modify encrypted data unless you know the secret.
+See [JWT](https://jwt.io/) for more info.
```
Comment (JerryHue) [inline src/web/docusaurus/docs/api-services/auth.md:72]:
```suggestion
See [JWT](https://jwt.io/introduction) for more info.
```

File: src/web/docusaurus/docs/api-services/auth.md [hunk 10]
```diff
@@ -1,3 +1,86 @@
...
+   :::
+
+## Proceeding stages
+
+- SSO service signs the token using a JWT secret, making it viable for a limited amount of time.
+
+:::info
+
+JWT is a string that has 3 parts. These parts are JavaScript objects made into a single line.
+Each part then base 64 encoded and is meant to be put in a URL. Parts are divided with `.` (dots)
```
Comment (JerryHue) [inline src/web/docusaurus/docs/api-services/auth.md:69]:
```suggestion
JWT is a string that is made up of 3 parts, joined by periods (`.`). Each part is a [`base64url`](https://www.rfc-editor.org/rfc/rfc4648#section-5) string. The original string is a JSON string, transformed through a `base64url` encoder.
```

**Review by JerryHue — CHANGES_REQUESTED** [review]

**Review by JiaHua-Zou — APPROVED** [review]

File: src/web/docusaurus/docs/api-services/auth.md [hunk 11]
```diff
@@ -1,3 +1,85 @@
...
+- Microsoft SAML-2.
+
+  :::info
+
+  See more in [Not so simple SAML](https://blog.humphd.org/not-so-simple-saml/)
+
+  :::
+
+- Auth cookies are returned to SSO service.
+SSO service creates a JWT token, which is basically encrypted JSON data.
```
Comment (JerryHue) [inline src/web/docusaurus/docs/api-services/auth.md:26]:
```suggestion
- Auth cookies are returned to SSO service.
  SSO service creates a JWT token, which is basically encrypted JSON data.
```

**Review by JiaHua-Zou — DISMISSED** [review]

**Review by JerryHue — APPROVED** [review]

**Review by AmasiaNalbandian — APPROVED** [review]

**Review by DukeManh — APPROVED** [review]
